File: packages/studio/src/api/lineage.ts
import { apiClient as api } from './client';

export interface Dataset {
  id: string;
  workspace_id: string;
  environment_id?: string;
  name: string;
  type: string;
  description?: string;
  owner?: string;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface LineageNode {
  id: string;
  type: string;
  name: string;
  data?: any;
}

export interface LineageEdge {
  upstream_id: string;
  downstream_id: string;
  pipeline_id?: string;
  execution_id?: string;
  confidence_level: string;
}

export interface LineageGraph {
  nodes: LineageNode[];
  edges: LineageEdge[];
}

export const lineageApi = {
  // Datasets
  getDatasets: async (): Promise<Dataset[]> => {
    const response = await api.get('/datasets');
    return response.data;
  },
  
  getDataset: async (id: string): Promise<Dataset> => {
    const response = await api.get(`/datasets/${id}`);
    return response.data;
  },
  
  createDataset: async (data: Partial<Dataset>): Promise<Dataset> => {
    const response = await api.post('/datasets', data);
    return response.data;
  },
  
  // Lineage
  getDatasetLineage: async (id: string): Promise<LineageGraph> => {
    const response = await api.get(`/lineage/${id}`);
    return response.data;
  },
  
  getDatasetImpact: async (id: string): Promise<any> => {
    const response = await api.get(`/lineage/impact/${id}`);
    return response.data;
  }
};

File: packages/server/flowcore_server/repositories/mappers/lineage.py
from typing import List
from flowcore_shared.schemas.lineage.dataset import Dataset, DatasetType
from flowcore_shared.schemas.lineage.graph import LineageEdge, ConfidenceLevel
from flowcore_server.db.lineage_models import Dataset as OrmDataset, LineageEdge as OrmLineageEdge
import uuid

def map_orm_to_dataset(orm_obj: OrmDataset) -> Dataset:
    return Dataset(
        id=str(orm_obj.id),
        workspace_id=str(orm_obj.workspace_id),
        environment_id=str(orm_obj.environment_id) if orm_obj.environment_id else None,
        name=orm_obj.name,
        type=DatasetType(orm_obj.type),
        description=orm_obj.description,
        owner=orm_obj.owner,
        version=orm_obj.version,
        created_at=orm_obj.created_at,
        updated_at=orm_obj.updated_at
    )

def map_orm_to_lineage_edge(orm_obj: OrmLineageEdge) -> LineageEdge:
    return LineageEdge(
        id=str(orm_obj.id),
        upstream_id=str(orm_obj.upstream_id),
        downstream_id=str(orm_obj.downstream_id),
        pipeline_id=str(orm_obj.pipeline_id) if orm_obj.pipeline_id else None,
        execution_id=str(orm_obj.execution_id) if orm_obj.execution_id else None,
        confidence_level=ConfidenceLevel(orm_obj.confidence_level),
        created_at=orm_obj.created_at
    )
